test/e2e/set_delete_test.go

```diff
@@ -0,0 +1,252 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+	appsv1 "k8s.io/api/apps/v1"
+	"strings"
+	"testing"
+	"time"
+
+	routev1 "github.com/openshift/api/route/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+var podCount int
+
+func testHeaders(t *testing.T, image string, route *routev1.Route, address string, headers []string, expectedResponse string, expectedMatches int) {
+	t.Helper()
+
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		t.Fatalf("failed to get kube config: %v", err)
+	}
+	client, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		t.Fatalf("failed to create kube client: %v", err)
+	}
+
+	var extraCurlArgs []string
+	for _, header := range headers {
+		extraCurlArgs = append(extraCurlArgs, "-H", header)
+	}
+	extraCurlArgs = append(extraCurlArgs, "-v", "--resolve", route.Spec.Host+":80:"+address)
+	podCount++
+	name := fmt.Sprintf("%s%d", route.Name, podCount)
+	clientPod := buildCurlPod(name, route.Namespace, image, route.Spec.Host, address, extraCurlArgs...)
+	if err := kclient.Create(context.TODO(), clientPod); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), clientPod); err != nil {
+			if !errors.IsNotFound(err) {
+				t.Fatalf("failed to delete pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+			}
+		}
+	}()
+	expectedResponse = strings.ToLower(expectedResponse)
+	err = wait.PollImmediate(1*time.Second, 4*time.Minute, func() (bool, error) {
+		readCloser, err := client.CoreV1().Pods(clientPod.Namespace).GetLogs(clientPod.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).Stream(context.TODO())
+		if err != nil {
+			t.Logf("failed to read output from pod %s: %v", clientPod.Name, err)
+			return false, nil
+		}
+		scanner := bufio.NewScanner(readCloser)
+		defer func() {
+			if err := readCloser.Close(); err != nil {
+				t.Errorf("failed to close reader for pod %s: %v", clientPod.Name, err)
+			}
+		}()
+		var numMatches int
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.Contains(strings.ToLower(line), expectedResponse) {
+				numMatches++
+				t.Logf("found match %d of %d expected: %s", numMatches, expectedMatches, line)
+			}
+		}
+		if numMatches > 0 && numMatches != expectedMatches {
```

Comment:
If numMatches == 0, then that could be an error for which we want to get an error message.
```suggestion
		if numMatches != expectedMatches {
```

Reply:
OK.

Reply:
without `if numMatches == 0`
the return gets executed instantly without even waiting and trying to match
`
[miheer@localhost cluster-ingress-operator]$ make TEST=TestSetRouteResponseHeaders test-e2e
go generate ./pkg/manifests
CGO_ENABLED=1 GO111MODULE=on GOFLAGS=-mod=vendor go test -timeout 1h -count 1 -v -tags e2e -run "TestSetRouteResponseHeaders" ./test/e2e
=== RUN   TestSetRouteResponseHeaders
=== PAUSE TestSetRouteResponseHeaders
=== CONT  TestSetRouteResponseHeaders
    set_delete_test.go:69: failed to read output from pod set-response-header-via-route-echo1: container "curl" in pod "set-response-header-via-route-echo1" is waiting to start: ContainerCreating
    set_delete_test.go:234: failed to observe the expected output: got 0 matches for "x-frame-options:", expected 1
    panic.go:522: deleted ingresscontroller set-response-headers-route
--- FAIL: TestSetRouteResponseHeaders (85.99s)
FAIL
FAIL	github.com/openshift/cluster-ingress-operator/test/e2e	88.409s
FAIL
make: *** [Makefile:56: test-e2e] Error 1
`

Reply:
with `if numMatches > 0 && numMatches != expectedMatches {`

the output:
`
[miheer@localhost cluster-ingress-operator]$ make TEST=TestSetRouteResponseHeaders test-e2e
go generate ./pkg/manifests
CGO_ENABLED=1 GO111MODULE=on GOFLAGS=-mod=vendor go test -timeout 1h -count 1 -v -tags e2e -run "TestSetRouteResponseHeaders" ./test/e2e
=== RUN   TestSetRouteResponseHeaders
=== PAUSE TestSetRouteResponseHeaders
=== CONT  TestSetRouteResponseHeaders
    set_delete_test.go:69: failed to read output from pod set-response-header-via-route-echo1: container "curl" in pod "set-response-header-via-route-echo1" is waiting to start: ContainerCreating
    set_delete_test.go:83: found match 1 of 1 expected: < x-frame-options: DENY
    set_delete_test.go:235: deleted ingresscontroller set-response-headers-route
--- PASS: TestSetRouteResponseHeaders (91.13s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/test/e2e	93.831s
[miheer@localhost cluster-ingress-operator]$ 

`

Reply:
I don't think we need this block at all then, if we wait and poll for the right number of matches to come up.

Reply:
See https://github.com/openshift/cluster-ingress-operator/pull/872/files#r1290482339